### PR TITLE
ramips: uboot-envtools: mtd: add support for Linksys EA8100

### PR DESCRIPTION
**Pull request by angelsl**
Add support for the Linksys EA8100.

With thanks to Leon Poon (@LeonPoon) for the [initial bringup](https://github.com/LeonPoon/openwrt/compare/master...master-linksys_ea8100).

### CODE REVIEW

File: target/linux/ramips/base-files/lib/ramips.sh [hunk 1]
```diff
@@ -415,6 +415,9 @@ ramips_board_detect() {
 	*"RE350 v1")
 		name="re350-v1"
 		;;
+	*"EA8100")
+		name="linksys,ea8100"
+		;;
```
Comment (adrianschmutzler) [inline target/linux/ramips/base-files/lib/ramips.sh]: This can be removed, as it will default to DTS compatible.

Comment (angelsl) [reply target/linux/ramips/base-files/lib/ramips.sh]: Done! Thank you.

**Comment by ynezz** [comment]
You're adding and changing a lot of files, it's hard to review, please split the changes into separate commits.

File: package/boot/uboot-envtools/files/ramips [hunk 2]
```diff
@@ -31,6 +31,7 @@ wsr-600|\
 zbt-wg2626)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x10000"
 	;;
+linksys,ea8100|\
```
Comment (ynezz) [inline package/boot/uboot-envtools/files/ramips]: Please make this separate commit `uboot-envtools: add support for Linksys EA8100`

**Comment by adrianschmutzler** [comment]
There has been a reorganization in ramips target today. Please:
- update your board.d subfiles since they have been re-sorted
- update your Makefile (mt****/rt****.mk) changes so sorting of devices is correct
- rename your DTS/DTSI files following the scheme soc_vendor_device.dts(i), e.g. mt7620a_tplink_archer-c50-v1.dts (and update includes there if necessary)
- remove `DTS := ...` from Makefile
- add `MTK_SOC` to Makefile (needs to match "soc" in DTS file name), e.g. `MTK_SOC := mt7620a`

Thank you!

For this PR, you might also have to address bootcount file ...

**Comment by angelsl** [comment]
Thanks!

I'm currently on vacation so I'll fix everything when I'm back in slightly more than a week.

**Comment by angelsl** [comment]
I've split the commits and also rebased on current master (including the reorganisation).

Let me know if I should split the PR too.

**Comment by LeonPoon** [comment]
Hey @angelsl. Thanks for putting the PR together.

Would recommend to use [mtdsplit](https://openwrt.org/docs/techref/flash.layout#mtdsplit) to scan for the rootfs, instead of hard coding the `rootfs` offset in dts.

If there is bad eraseblock in the nand in the kernel partition (such as in my case) then flashing the openwrt image will shift everything after the bad block by 1. Hence the `rootfs` in dts will point to actually kernel partition data. The kernel will fail to mount rootfs because there won't be squashfs/ubifs signature at `rootfs`.

Maybe you can use [my dts](https://github.com/LeonPoon/openwrt/blob/master-linksys_ea8100/target/linux/ramips/dts/EA8100.dts) and [this commit](https://github.com/LeonPoon/openwrt/commit/a122e68ff45aba6a85ea7d2ab0b773a5e773a4c6#diff-5476a825952c25a3c92968c68aeabce7) for reference.

File: target/linux/ramips/dts/mt7621_linksys_ea8100.dts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	};
+
+	partition@140000 {
+		label = "devinfo";
+		reg = <0x140000 0x40000>;
+	};
+
+	partition@180000 {
+		label = "kernel1";
+		reg = <0x180000 0x2800000>;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_linksys_ea8100.dts]: Size and position of the next partition do not match ...?

File: target/linux/ramips/dts/mt7621_linksys_ea8100.dts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	};
+
+	partition@580000 {
+		label = "rootfs1";
+		reg = <0x580000 0x2400000>;
+	};
+
+	partition@2980000 {
+		label = "kernel2";
+		reg = <0x2980000 0x2800000>;
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_linksys_ea8100.dts]: Size and position of the next partition do not match ...?

File: target/linux/ramips/dts/mt7621_linksys_ea8100.dts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+		label = "sysdiag";
+		reg = <0x5180000 0x100000>;
+		read-only;
+	};
+
+	partition@5280000 {
+		label = "syscfg";
+		reg = <0x5280000 0x2d00000>;
+		read-only;
+	};
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_linksys_ea8100.dts]: 0x80000 missing to 0x8000000. By intention?

**Comment by angelsl** [comment]
@adrianschmutzler The kernel1/kernel2 partitions encompass the rootfs1/rootfs2 partitions. I based this on [how the Linksys EA8500's partition layout was done](https://github.com/openwrt/openwrt/blob/master/target/linux/ipq806x/files-4.14/arch/arm/boot/dts/qcom-ipq8064-ea8500.dts#L275), since the EA8500 is also a NAND device that has dual firmware partitions.

As for the syscfg partition, that is based on the partition information from the factory firmware.

@LeonPoon Yeah, I saw that. But mtdsplit as far as I can tell can't be controlled by kernel arguments currently, so I'd have to write code to handle dual-booting. I was trying to minimise the amount of lines of kernel patch, and also I just followed how the EA8500 did things (since it also has a NAND). In fact `upgrade.sh` was lifted from [the EA8500](https://github.com/openwrt/openwrt/blob/master/target/linux/ipq806x/base-files/lib/upgrade/linksys.sh).

**Comment by angelsl** [comment]
For reference, from the factory firmware:

```
/proc # cat mtd
dev:    size   erasesize  name
mtd0: 07f80000 00020000 "ALL"
mtd1: 00080000 00020000 "boot"
mtd2: 00040000 00020000 "u_env"
mtd3: 00040000 00020000 "Factory"
mtd4: 00040000 00020000 "s_env"
mtd5: 00040000 00020000 "devinfo"
mtd6: 02800000 00020000 "kernel"
mtd7: 02400000 00020000 "rootfs"
mtd8: 02800000 00020000 "alt_kernel"
mtd9: 02400000 00020000 "alt_rootfs"
mtd10: 00100000 00020000 "sysdiag"
mtd11: 02d00000 00020000 "syscfg"
```

**Comment by adrianschmutzler** [comment]
@angelsl Okay, looks like this is common practice. Also found two cases in mvebu where they added comments:
https://github.com/openwrt/openwrt/blob/016d1eb1f9c14c50dbd0b5461fc22c1f4ef6e30a/target/linux/mvebu/files-4.19/arch/arm/boot/dts/armada-385-linksys-venom.dts#L154
I'd personally add the comments in DTS, but I've seen reviewers discouraging comments in DTS recently, so I am not sure on recommending this.

**Comment by angelsl** [comment]
I just remembered that [@LeonPoon wrote a patch to mtdsplit](https://github.com/LeonPoon/openwrt/commit/b90086cc98a15a2774382b691f1249f4ad11a1c7) to choose the partition to split based on the arguments from the bootloader. I chose to follow how other already-supported devices handled NAND with dual-boot instead.

But given what @LeonPoon said:

> If there is bad eraseblock in the nand in the kernel partition (such as in my case) then flashing the openwrt image will shift everything after the bad block by 1. Hence the `rootfs` in dts will point to actually kernel partition data. The kernel will fail to mount rootfs because there won't be squashfs/ubifs signature at `rootfs`.

It seems like the other dual-booting NAND devices just ignored this issue? (e.g. EA8500)

**Comment by angelsl** [comment]
[This issue](https://bugs.openwrt.org/index.php?do=details&task_id=1926) seems relevant.

AIUI, the current MT7621 NAND driver remaps bad blocks transparently, so it should be okay (?) to have a fixed partition layout. The question is does the factory firmware and bootloader remap blocks in the same way? If not, a flash from factory and/or U-Boot might place the rootfs somewhere the kernel doesn't expect.

File: target/linux/ramips/image/mt7621.mk [hunk 6]
```diff
@@ -293,6 +293,22 @@ define Device/lenovo_newifi-d1
 endef
 TARGET_DEVICES += lenovo_newifi-d1
 
+define Device/linksys_ea8100
+  MTK_SOC := mt7621
```
Comment (adrianschmutzler) [inline target/linux/ramips/image/mt7621.mk]: use SOC instead.

File: target/linux/ramips/image/mt7621.mk [hunk 7]
```diff
@@ -293,6 +293,22 @@ define Device/lenovo_newifi-d1
...
+  PAGESIZE := 2048
+  UBINIZE_OPTS := -E 5
+  BLOCKSIZE := 128k
+  FILESYSTEMS := squashfs
+  DEVICE_TITLE := Linksys EA8100
```
Comment (adrianschmutzler) [inline target/linux/ramips/image/mt7621.mk]: Split into DEVICE_MODEL/DEVICE_VENDOR

File: target/linux/ramips/image/mt7621.mk [hunk 8]
```diff
@@ -293,6 +293,22 @@ define Device/lenovo_newifi-d1
...
+  DEVICE_PACKAGES := kmod-mt7615e wpad-basic kmod-usb3 uboot-envtools kmod-usb-storage
+  IMAGE_SIZE := 40960k
+  KERNEL_SIZE := 4096k
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  IMAGE/factory.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=EA8100
```
Comment (adrianschmutzler) [inline target/linux/ramips/image/mt7621.mk]: Wrap line after 80 chars.

File: target/linux/ramips/dts/mt7621_linksys_ea8100.dts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
```
Comment (adrianschmutzler) [inline target/linux/ramips/dts/mt7621_linksys_ea8100.dts]: You can use state_default directly,
```
&state_default {
	gpio {
		ralink,group = "i2c", "uart2", "uart3", "jtag", "wdt";
		ralink,function = "gpio";
	};
};
```

**Comment by angelsl** [comment]
Sorry, I don't have the time to fix this PR now.

If someone could do it, that'd be great.

**Comment by adrianschmutzler** [comment]
Since this PR is quite old, and the author stated he won't be able to update it, I'm closing it now.

As it's targetting mt7621, where a lot of changes happened in the meantime, I think resubmitting it with all the feedback implemented might be the best choice if somebody wants to continue.
